### ISSUE TITLE
Need clean target?

### ISSUE THREAD

**Issue by joseogl** [MicrosoftDocs/visualstudio-docs#4290]
I am gettimg errors when using Build.Clean or Build.Rebuild.

------ Clean started: Project: NodejsWebApp1, Configuration: Debug Any CPU ------
		npm ERR! missing script: clean
		npm ERR! A complete log of this run can be found in:
		npm ERR!     C:\Users\joseo\AppData\Roaming\npm-cache\_logs\2019-11-07T15_16_15_572Z-debug.log
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VisualStudio\v16.0\Node.js Tools\Microsoft.NodejsToolsV2.targets(56,5): error MSB3073: The command "npm run clean" exited with code 1.
Done building project "NodejsWebApp1.njsproj" -- FAILED.

Build FAILED.
========== Clean: 0 succeeded, 1 failed, 0 sk
[Enter feedback here]

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: c4b0ea28-4a31-a3ea-fa5f-c32d5c72d4bd
* Version Independent ID: 31bd064b-51eb-5505-986e-2c7148784085
* Content: [Create a Node.js and React app - Visual Studio](https://docs.microsoft.com/en-us/visualstudio/javascript/tutorial-nodejs-with-react-and-jsx?view=vs-2019#feedback)
* Content Source: [docs/javascript/tutorial-nodejs-with-react-and-jsx.md](https://github.com/MicrosoftDocs/visualstudio-docs/blob/master/docs/javascript/tutorial-nodejs-with-react-and-jsx.md)
* Product: **visual-studio-windows**
* Technology: **vs-javascript**
* GitHub Login: @Mikejo5000
* Microsoft Alias: **mikejo**

**Comment by Mikejo5000** [comment]
@joseogl Hi, at what section and iinstruction did you run into this error? Do you run into errors when not using Clean or Rebuild and just following the instructions as described in the article? Thx!

**Comment by Mikejo5000** [comment]
Some of the templates previously had a post build event in the .njsproj file, like this, which supported the Clean/Rebuild:

<PostBuildEvent>npm run build</PostBuildEvent>

See this:
https://docs.microsoft.com/en-us/visualstudio/javascript/quickstart-vuejs-with-nodejs?view=vs-2019#build-the-project

I believe this element was removed from more recent projects for some reason.

**Comment by joseogl** [comment]
Thanks.  I just got this error while trying to rebuild while exploring the project. I am trying to get debugging the client side to work without any luck do far.

**Comment by Mikejo5000** [comment]
What version of Visual Studio are you using? And what browser are you running in?

**Comment by Mikejo5000** [comment]
@armanio123 Armando, do you know why Clean/Rebuild doesn't work in the templates? Looks like it might be due to a template change, as I noted above.

**Comment by joseogl** [comment]
I am using VS2019 Community v 16.3.8 with Edge, Chrome, and Edge Beta. No matter what, I can't breakpoints to hit on the client code.  Server side code hits breakpoints.

**Comment by Mikejo5000** [comment]
Did you see any errors? What was the result? It is a tricky area with source maps and transpiled code. I know that the steps to do this need to be exact:
1. Set a breakpoint in client-side code in an area of code that will execute.
2. Make sure your sourcemaps are correct. Both relative path and prefixes. In this example, the array should have a reference to "./app.tsx" and not "webpack:///./app.tsx". If you've moved around source files, that can be a source of path issues, too.
3. Close all browser instances (use Task Manager to close extensions).
4. Open the browser in debug mode. I suggest using the Start > Run command line if you tried the UI config already.
5. Start **without** debugging.
6. Attach to process.
7. In Attach to Process dialog box, select the correct script debugger type, as described, and Attach to the correct browser instance.
8. Restart the browser.

If these steps don't work, you might need to use the F12 tools for client-side code.

**Comment by Mikejo5000** [comment]
Also, with Edge Beta (Chromium base), I believe I've seen Chrome extensions interfere with getting the Edge Beta browser into debug mode.

**Comment by armanio123** [comment]
Visual Studio uses the `build` and `clean` scripts of the `package.json` when executing either Build, Rebuild or Clean . The error message states that the `clean` script is missing. To fix it you can either add it to the `package.json` or remove the command my modifying the project.

Examples:
Add to `package.json` (assuming you're using TypeScript):
```json
...
scripts {
  "build": "tsc --build",
  "clean": "tsc --build --clean"
}
...
```

To remove the commands all together, add the following lines to your project (right click > unload project, right click > edit project) inside the project node:
```xml
<Project DefaultTargets="Build" ... >
...
  <Target Name="AfterClean"></Target>
  <Target Name="AfterCompile"></Target>
</Project>
```

The scripts might be tailored based on the your project kind and needs.

**Comment by joseogl** [comment]
Thanks @armanio123. I will try the package.json changes.

@Mikejo5000 I have followed the steps for debugging as suggested.  Here are my results.

I was unable to launch Microsoft Edge (not the Beta) with the debugger port argument suggested. U switched to Chrome and that allowed me to set the breakpoint in the JS file not TS file (because of the change to webpack-config.js?). Unfortunately I get a message in VS saying that a debugger operation is taking too long to complete. My breakpoint flashes and Chrome then displays a message saying that the script is Paused I Visual Studio Code.  (I do have Visual Studio Code but am using Visual Studio).  Upon hitting F5 in Visual Studio, the text renders but I don't get any more breakpoint hits. So closer but still no cigar. Thanks for the help.

**Comment by Mikejo5000** [comment]
Yeah, I think only Chrome or the Edge Beta will work for debugging with source maps. I don't think regular Edge can do it.

**Comment by Mikejo5000** [comment]
@joseogl BTW, I have seen that exact issue with script paused in Visual Studio Code. If I recall correctly, the issue was either the wrong relative path in the source map to "webpack:///...." or it was that the browser was not getting into debug mode properly, with other instances of Edge (Chromium) / Chrome / Chrome extensions interfering.

Besides F12 tools, another option would be to try the project with Visual Studio Code. They have a debugging extension for Chrome and Edge (Chromium) that may simplify this whole scenario.

**Comment by WilliamAntonRohm** [comment]
I am closing this issue now.  You are welcome to @ mention me for any followup.
We hope to hear from you again.

**Comment by Mikejo5000** [comment]
Keeping this open for upcoming doc on compile and build. Some of the comments above may be helpful.

**Comment by Mikejo5000** [comment]
Adding docs on compile and build.
